amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+            let (token_out_amount, swap_fee) = self._swap_to(
+                token_in_id,
+                token_out_id,
+                token_in_amount,
+                min_token_out_amount,
+            )?;
+
+            // transfer token_out
+            self.token_by_address(token_out)
+                .transfer(to, token_out_amount, vec![])?;
+
+            self.env().emit_event(Swap {
+                sender: self.env().caller(),
+                token_in,
+                amount_in: token_in_amount,
+                token_out,
+                amount_out: token_out_amount,
+                to,
+            });
+            Ok((token_out_amount, swap_fee))
+        }
+
+        #[ink(message)]
+        fn swap_exact_out(
+            &mut self,
+            token_in: AccountId,
+            token_out: AccountId,
+            token_out_amount: u128,
+            max_token_in_amount: u128,
+            to: AccountId,
+        ) -> Result<(u128, u128), StablePoolError> {
+            // Make sure rates are up to date before we attempt any calculations
+            self.update_rates();
+
+            //check token ids
+            let (token_in_id, token_out_id) = self.check_tokens(token_in, token_out)?;
+
+            // calculate amount out, mint admin fee and update reserves
+            let (token_in_amount, swap_fee) = self._swap_from(
+                token_in_id,
+                token_out_id,
+                token_out_amount,
+                max_token_in_amount,
+            )?;
+
+            // transfer token_in
+            self.token_by_address(token_in).transfer_from(
+                self.env().caller(),
+                self.env().account_id(),
+                token_in_amount,
+                vec![],
+            )?;
+
+            // transfer token_out
+            self.token_by_address(token_out)
+                .transfer(to, token_out_amount, vec![])?;
+
+            self.env().emit_event(Swap {
+                sender: self.env().caller(),
+                token_in,
+                amount_in: token_in_amount,
+                token_out,
+                amount_out: token_out_amount,
+                to,
+            });
+            Ok((token_in_amount, swap_fee))
+        }
+
+        #[ink(message)]
+        fn set_owner(&mut self, new_owner: AccountId) -> Result<(), StablePoolError> {
+            self.ensure_onwer()?;
+            self.owner = new_owner;
+            Ok(())
+        }
+
+        #[ink(message)]
+        fn set_fee_receiver(
+            &mut self,
+            fee_receiver: Option<AccountId>,
+        ) -> Result<(), StablePoolError> {
+            self.ensure_onwer()?;
+            self.pool.fee_receiver = fee_receiver;
+            Ok(())
+        }
+
+        #[ink(message)]
+        fn set_amp_coef(&mut self, amp_coef: u128) -> Result<(), StablePoolError> {
+            self.ensure_onwer()?;
+            validate_amp_coef(amp_coef)?;
+            self.pool.amp_coef = amp_coef;
```

Comment:
If we want to emit an event on changing AmpCoeff - it should happen here.

Reply:
Added events https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/4/commits/60e18d5f5b55f4cb6269b3b397943a3de9a803c1